syscall/Makefile

```diff
@@ -76,27 +88,43 @@ $(BIN1): $(PROXY_OBJS)
 $(BIN2): $(STUB_OBJS)
 	$(call ARCHIVE, $@, $(STUB_OBJS))
 
+$(BIN3): $(WRAP_OBJS)
+	$(call ARCHIVE, $@, $(WRAP_OBJS))
+
+$(SYSCALLIST): .context
+
 .depend: Makefile $(SRCS)
-	$(Q) $(MKDEP) $(PROXYDEPPATH) $(STUBDEPPATH) \
+	$(Q) $(MKDEP) $(PROXYDEPPATH) $(STUBDEPPATH) $(WRAPDEPPATH) \
 	  "$(CC)" -- $(CFLAGS) -- $(SRCS) >Make.dep
 	$(Q) touch $@
 
 depend: .depend
 
 .context: syscall.csv
 	$(Q) $(MAKE) -C $(TOPDIR)$(DELIM)tools -f Makefile.host mksyscall
+ifeq ($(CONFIG_LIB_SYSCALL),y)
 	$(Q) (cd proxies; $(MKSYSCALL) -p $(CSVFILE);)
 	$(Q) (cd stubs; $(MKSYSCALL) -s $(CSVFILE);)
+else
+ifeq ($(CONFIG_SCHED_INSTRUMENTATION_SYSCALL),y)
+	$(Q) (cd wraps; $(MKSYSCALL) -w $(CSVFILE);)
+	$(Q) $(MKSYSCALL) -l $(CSVFILE) > $(SYSCALLLIST:.txt=.h)
+	$(Q) $(call PREPROCESS, $(SYSCALLLIST:.txt=.h), $(SYSCALLLIST))
+	$(Q) sed -i -n -e '/^[^#]/p' $(SYSCALLLIST)
```

Comment:
Why we need four steps to generate the link command file?
```
$(Q) $(MKSYSCALL) -l $(CSVFILE) > $(SYSCALLLIST:.txt=.h)
$(Q) $(call PREPROCESS, $(SYSCALLLIST:.txt=.h), $(SYSCALLLIST))
$(Q) sed -i -n -e '/^[^#]/p' $(SYSCALLLIST)
cat $< | sed -e 's/^/--wrap=/' > $@
```
We can define a clever SYSCALL_LOOKUP macro to generate them in one step:
```
--wrap=xxx
--wrap=yyy
```
then we don't add -l for mksyscall and postprocess again in arch's Makefile.

Reply:
Because of the post process in arch's Makefile is that the '--wrap' option depends on GNU linker and I didn't want to add such code into arch independent syscall/ directory. 
As you commented, the first three steps seems to be able to realize by using SYSCALL_LOOKUP macro without mksyscall -l.

How about this?
- Revert mksyscall change for -l option and fix syscall/Makefile not to use it.
- No change in arch's Makefile

Reply:
The suggestion is reasonable to support the different toolchain, or we can let arch ovevwrite the macro expansion process like this:
```
#include <nuttx/config.h>
#include <syscall.h>
#include <arch.h>

#ifndef UP_WRAP
#define UP_WRAP(f) --wrap=##f
#endif

define SYSCALL_LOOKUP1(f,n) UP_WRAP(f)
define SYSCALL_LOOKUP(f,n) UP_WRAP(f)
#include <sys/syscall_lookup.h>
#undef SYSCALL_LOOKUP1
#undef SYSCALL_LOOKUP
```

Reply:
I tried to take your way and found that cpp generates the error because `--wrap=xxx` is invalid as C tokens.
It seems should be avoided to handle non-C-source text directly by C preprocessor.
So I suggests the alternative idea to prepare the host command to generate the command line for linker such as:
```
#include <stdio.h>
#include <nuttx/config.h>
#include <arch/arch.h>

#ifndef UP_WRAPOPT
#define UP_WRAPOPT(f) "--wrap=" #f " "
#endif

int main()
{
    printf("%s",
#define SYSCALL_LOOKUP1(f,n) UP_WRAPOPT(f)
#define SYSCALL_LOOKUP(f,n)  UP_WRAPOPT(f)
#include <sys/syscall_lookup.h>
#undef SYSCALL_LOOKUP1
#undef SYSCALL_LOOKUP
    );
}
```
It is built and executed by the host environment.
In my brief trial it seems work as my expectation, but my concern is whether the host tool is permitted to include "nuttx/config.h" and "arch/arch.h" directly.
Are there available for such purpose? Or do you know better way for it?

BTW, I found that similar consideration should be needed for wrapper file because `__wrap_` and `__real_` also depend on GNU linker.
I'll take similar way with mksyscall to generate the wrapper code such as:
```
#include <nuttx/arch.h>
#ifndef UP_WRAPSYM
#define UP_WRAPSYM(s) __wrap_##s
#endif
#ifndef UP_REALSYM
#define UP_REALSYM(s) __real_##s
#endif

int UP_WRAPSYM(open)(FAR const char * parm1, int parm2, ...)
  :
```
Thanks for good suggestion.

Reply:
cpp is a general tool which can handle any text file, for example:
1.This commit process nuttx-names.dat by cpp:
https://github.com/apache/incubator-nuttx/pull/1262/commits/abf77a6dc1bec20b38df2ae2d37bed6109ad125c
2.This commit process shell script by cpp:
https://github.com/apache/incubator-nuttx/pull/793/commits/925d0107451dd50b8637f2a07f1dceb2b90d09d2
The key point is that we have to stop the normal step(compile) after preprocess.

Reply:
cpp ## operator cannot concatenate arbitrary string.
I found that the C language standard declares that both arguments of the ## operator and the result of the concatenation must be valid preprocessing tokens.
If not, the behavior is undefined in the standard.

In fact, if passing the following code into GNU cpp,
```
#define CONCAT_(x, y) x##y
#define CONCAT(x, y)  CONCAT_(x, y)

CONCAT(wrap, xxx)
CONCAT(--wrap=, xxx)
```
The former CONCAT works well but the latter gets the error `error: pasting "=" and "xxx" does not give a valid preprocessing token`.

In your example, abf77a6 concatenates the correct tokens.
925d010 concatenates `CONCAT(/dev/ram, CONFIG_NSH_FATDEVNO)`, works because `CONFIG_NSH_FATDEVNO` has numeric value and the result like "ram0" is the correct token.
(But I think "/dev/ram0" is still incorrect token... Strictly speaking, it seems undefined result in the standard.)

Anyway, in our case, because "--wrap=" is not correct token, I think that we cannot take this way unfortunately.

Reply:
Oops. Just using "--wrap xxx" instead of "--wrap=xxx" solves it....
GNU linker also accepts the former styles.
I'll try it again, sorry.